### PR TITLE
Add ZenMLServiceType and update service registration

### PR DESCRIPTION
**Pull request by safoinme**
## Describe changes
I implemented/fixed _ to achieve _.

## Pre-requisites
Please ensure you have done the following:
- [x] I have read the **CONTRIBUTING.md** document.
- [x] If my change requires a change to docs, I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] I have based my new branch on `develop` and the open PR is targeting `develop`. If your branch wasn't based on develop read [Contribution guide on rebasing branch to develop](https://github.com/zenml-io/zenml/blob/main/CONTRIBUTING.md#-pull-requests-rebase-your-branch-on-develop).
- [ ] If my changes require changes to the dashboard, these changes are communicated/requested.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Other (add details above)

## Summary by CodeRabbit

- **New Features**
    - Enhanced artifact loading process by adding functionality to load an artifact from a response using a client and its ID.
    - Improved CLI functionality by adding options to filter served model versions by `model-version` and `flavor`.
    - Updated service handling methods in the ZenML client for creating, getting, listing, updating, and deleting services.
    - Modified constants and methods in the Bentoml integration for improved service deployment.
    - Added a `logo_url` field to the SeldonDeploymentService class for linking to the Seldon logo.
    - Introduced methods in various files for linking services to models and handling service instances.
    - Updated service status return values in multiple files for more accurate deployment status reporting.
    - Added new states and methods for managing service and service endpoint states.
    - Updated comparison checks in various files to reflect accurate service states.
    - Introduced new entities, methods, and functionalities for service management across different ZenML components.

### CODE REVIEW

**Comment by safoinme** [comment]
@coderabbitai review

**Comment by safoinme** [comment]
@coderabbitai review

**Comment by safoinme** [comment]
@coderabbitai review

**Review by avishniakov — CHANGES_REQUESTED** [review]
Great one, left a few comments.
Mosty important: 
- it is not clear how the new service field of the Model will cooperate with existing logic.
- no tests are created

File: src/zenml/constants.py [hunk 1]
```diff
@@ -230,6 +230,13 @@ def handle_int_env_var(var: str, default: int = 0) -> int:
 SERVICE_CONNECTOR_RESOURCES = "/resources"
 SERVICE_CONNECTOR_TYPES = "/service_connector_types"
 SERVICE_CONNECTOR_VERIFY = "/verify"
+SERVICE_CONNECTOR_RESOURCES = "/resources"
+SERVICE_CONNECTOR_CLIENT = "/client"
+MODELS = "/models"
+MODEL_VERSIONS = "/model_versions"
+MODEL_VERSION_ARTIFACTS = "/model_version_artifacts"
+MODEL_VERSION_PIPELINE_RUNS = "/model_version_pipeline_runs"
```
[the remark this answers is not in the document]
Comment (safoinme) [reply src/zenml/constants.py:238]: I think this was just a merge issue

File: src/zenml/materializers/service_materializer.py [hunk 2]
```diff
@@ -14,20 +14,21 @@
...
-from zenml.services.service_registry import ServiceRegistry
+from zenml.services.service import BaseDeploymentService, BaseService
 
 if TYPE_CHECKING:
     from zenml.metadata.metadata_types import MetadataType
 
 SERVICE_CONFIG_FILENAME = "service.json"
 
 
+# TODO: refactor service materializer to not store into artifact store but get it from db
```
[the remark this answers is not in the document]
Comment (safoinme) [reply src/zenml/materializers/service_materializer.py:31]: that's was leftover, thank you

File: src/zenml/zen_stores/schemas/model_schemas.py [hunk 3]
```diff
@@ -263,6 +269,10 @@ class ModelVersionSchema(NamedSchema, table=True):
         ),
     )
 
+    services: List["ServiceSchema"] = Relationship(
+        back_populates="model_version",
+    )
```
[the remark this answers is not in the document]
Comment (safoinme) [reply src/zenml/zen_stores/schemas/model_schemas.py:274]: This is part of the newly introduced `resources response` which was part of the changes in the triggers PR. The goal is to eventually deprecate the  `get_deployment_artifact` as the service won't be considered as artifacts anymore

File: src/zenml/materializers/service_materializer.py [hunk 4]
```diff
@@ -64,7 +68,7 @@ def save(self, service: BaseService) -> None:
         artifact_store = Client().active_stack.artifact_store
         filepath = os.path.join(self.uri, SERVICE_CONFIG_FILENAME)
         with artifact_store.open(filepath, "w") as f:
-            f.write(service.json(indent=4))
+            f.write(str(service.uuid))
```
[the remark this answers is not in the document]
Comment (avishniakov) [reply src/zenml/materializers/service_materializer.py:71]: Yeah, makes sense - missed that you are saving just UUID

File: src/zenml/zen_stores/schemas/model_schemas.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (avishniakov) [reply src/zenml/zen_stores/schemas/model_schemas.py:274]: Ok, can you create a ticket for that, if we don't have one, so we keep it in focus?

File: docs/book/stacks-and-components/component-guide/model-deployers/model-deployers.md [hunk 5]
```diff
@@ -87,46 +82,52 @@ zenml model-deployer register seldon --flavor=seldon \
...
+`find_model_server` - Finds and returns a list of Service objects that represent model servers that have been deployed to the serving environment, the 
+services are stored in the DB and can be used as a reference to know what and where the model is deployed.
+`stop_model_server` - Stops a model server that is currently running in the serving environment.
+`start_model_server` - Starts a model server that has been stopped in the serving environment.
+`delete_model_server` - Deletes a model server from the serving environment and from the DB.
+
+{% hint style="info" %}
+ZenML uses the Service object to represent a model server that has been deployed to a serving environment. The Service object is saved in the DB and can be used as a reference to know what and where the model is deployed. The Service object consist of 2 main attributes, the `config` and the `status`. The `config` attribute holds all the deployment configuration attributes required to create a new deployment, while the `status` attribute holds the operational status of the deployment, such as the last error message, the prediction URL, and the deployment status.
```
Comment (bcdurak) [inline docs/book/stacks-and-components/component-guide/model-deployers/model-deployers.md:95]:
```suggestion
ZenML uses the Service object to represent a model server that has been deployed to a serving environment. The Service object is saved in the DB and can be used as a reference to know what and where the model is deployed. The Service object consists of 2 main attributes, the `config` and the `status`. The `config` attribute holds all the deployment configuration attributes required to create a new deployment, while the `status` attribute holds the operational status of the deployment, such as the last error message, the prediction URL, and the deployment status.
```

File: src/zenml/services/service.py [hunk 6]
```diff
@@ -476,22 +456,49 @@ def stop(self, timeout: int = 0, force: bool = False) -> None:
...
+        prediction_url = None
+        if isinstance(self, BaseDeploymentService) and self.prediction_url:
+            prediction_url = self.prediction_url
+        elif self.endpoint:
+            prediction_url = (
+                self.endpoint.status.uri if self.endpoint.status else None
+            )
+        return prediction_url
+
+    def get_healthcheck_url(self) -> Optional[str]:
```
Comment (bcdurak) [inline src/zenml/services/service.py:487]: There is a `get_healthcheck_url` method and a `healthcheck_url` property. Should they be somewhat related to each other?

**Comment by bcdurak** [comment]
I agree with @avishniakov here. Some additional tests could be really critical here.

File: src/zenml/models/v2/core/service.py [hunk 7]
```diff
@@ -0,0 +1,479 @@
...
+    model_version_id: Optional[Union[UUID, str]] = Field(
+        default=None,
+        description="By the model version this service is attached to.",
+    )
+    pipeline_run_id: Optional[Union[UUID, str]] = Field(
+        default=None,
+        description="By the pipeline run this service is attached to.",
+    )
+
+    def set_type(self, type: str) -> None:
```
[the remark this answers is not in the document]
Comment (safoinme) [reply src/zenml/models/v2/core/service.py:405]: The idea was to set it to call `zen_store.list_services` with type scoping by setting it explicitly as defined in `def get_service` in client

File: src/zenml/models/v2/core/service.py [hunk 8]
```diff
@@ -0,0 +1,479 @@
...
+        "running",
+        "pipeline_name",
+    ]
+
+    def generate_filter(
+        self, table: Type["SQLModel"]
+    ) -> Union["BinaryExpression[Any]", "BooleanClauseList[Any]"]:
+        """Generate the filter for the query.
+
+        Stack components can be scoped by type to narrow the search.
```
[the remark this answers is not in the document]
Comment (safoinme) [reply src/zenml/models/v2/core/service.py:447]: yes thank you for the catch

File: src/zenml/zen_stores/sql_zen_store.py [hunk 9]
```diff
@@ -1762,6 +1769,180 @@ def delete_api_key(
...
+                session=session,
+            )
+
+            # Create the service.
+            service_schema = ServiceSchema.from_request(service)
+            logger.debug("Creating service: %s", service_schema)
+            session.add(service_schema)
+            session.commit()
+
+            return service_schema.to_model(include_metadata=True)
```
[the remark this answers is not in the document]
Comment (safoinme) [reply src/zenml/zen_stores/sql_zen_store.py:1832]: Yes that's right, adding it now

File: src/zenml/zen_stores/schemas/service_schemas.py [hunk 10]
```diff
@@ -0,0 +1,252 @@
...
+                endpoint=json.loads(base64.b64decode(self.endpoint).decode())
+                if self.endpoint
+                else None,
+                admin_state=self.admin_state or None,
+                prediction_url=self.prediction_url or None,
+                health_check_url=self.health_check_url,
+            )
+        resources = None
+        if include_resources:
+            resources = ServiceResponseResources(
```
[the remark this answers is not in the document]
Comment (safoinme) [reply src/zenml/zen_stores/schemas/service_schemas.py:159]: I had it defined before, but the thing is that we don't need the service to return the pipeline run or model version response if hydrate is set to True, at least for the moment the service is only a tab within the model version and the actual response we want is just the service

File: src/zenml/services/service.py (hunk 6 above)
Comment (safoinme) [reply src/zenml/services/service.py:487]: the `healthcheck_url` is a property that can be set for the service itself, the `get_healthcheck_url` check the health check for the service or the endpoint and return the first one it finds defined

File: src/zenml/services/service.py [hunk 11]
```diff
@@ -26,24 +28,27 @@
...
 
 T = TypeVar("T", bound=Callable[..., Any])
 
+if TYPE_CHECKING:
+    from zenml.models.v2.core.service import ServiceResponse
+
+ZENM_ENDPOINT_PREFIX = "zenml-"
+
 
 def update_service_status(
```
[the remark this answers is not in the document]
Comment (safoinme) [reply src/zenml/services/service.py:53]: sorry can you elaborate more, i didn't understand the question?

File: docs/book/stacks-and-components/component-guide/model-deployers/model-deployers.md [hunk 12]
```diff
@@ -87,46 +82,52 @@ zenml model-deployer register seldon --flavor=seldon \
...
-   the remote model server, e.g. to get the operational status of a model server, the prediction URI that it exposes, or
-   to stop or delete a model server:
+* Lifecycle Management: Provides mechanisms for comprehensive lifecycle management of model servers, including the ability to start, stop, and delete model servers, as well as to update existing servers with new model versions, thereby optimizing resource utilization and facilitating continuous delivery of model updates. Some core methods that can be used to interact with the remote model server include:
+
+`deploy_model` - Deploys a model to the serving environment and returns a Service object that represents the deployed model server.
+`find_model_server` - Finds and returns a list of Service objects that represent model servers that have been deployed to the serving environment, the 
+services are stored in the DB and can be used as a reference to know what and where the model is deployed.
+`stop_model_server` - Stops a model server that is currently running in the serving environment.
+`start_model_server` - Starts a model server that has been stopped in the serving environment.
```
Comment (strickvl) [inline docs/book/stacks-and-components/component-guide/model-deployers/model-deployers.md:91]: When you start it for the first time, it's starting without being stopped, no?

File: docs/book/stacks-and-components/component-guide/model-deployers/model-deployers.md [hunk 13]
```diff
@@ -87,46 +82,52 @@ zenml model-deployer register seldon --flavor=seldon \
...
+`delete_model_server` - Deletes a model server from the serving environment and from the DB.
+
+{% hint style="info" %}
+ZenML uses the Service object to represent a model server that has been deployed to a serving environment. The Service object is saved in the DB and can be used as a reference to know what and where the model is deployed. The Service object consists of 2 main attributes, the `config` and the `status`. The `config` attribute holds all the deployment configuration attributes required to create a new deployment, while the `status` attribute holds the operational status of the deployment, such as the last error message, the prediction URL, and the deployment status.
```
Comment (strickvl) [inline docs/book/stacks-and-components/component-guide/model-deployers/model-deployers.md:95]: Wondering whether to give the details of where the service object is being saved... we don't talk about that elsewhere in the documentation. Wonder whether it's confusing for a user. Maybe better to say something like 'saved on the ZenML Server' or something, or 'saved as metadata on the ZenML server' etc?

File: docs/book/stacks-and-components/component-guide/model-deployers/model-deployers.md [hunk 14]
```diff
@@ -87,46 +82,52 @@ zenml model-deployer register seldon --flavor=seldon \
...
-   listed and filtered using a variety of criteria, such as the name of the model or the names of the pipeline and step
-   that was used to deploy the model. The Service objects returned by the Model Deployer can be used to interact with
-   the remote model server, e.g. to get the operational status of a model server, the prediction URI that it exposes, or
-   to stop or delete a model server:
+* Lifecycle Management: Provides mechanisms for comprehensive lifecycle management of model servers, including the ability to start, stop, and delete model servers, as well as to update existing servers with new model versions, thereby optimizing resource utilization and facilitating continuous delivery of model updates. Some core methods that can be used to interact with the remote model server include:
+
+`deploy_model` - Deploys a model to the serving environment and returns a Service object that represents the deployed model server.
+`find_model_server` - Finds and returns a list of Service objects that represent model servers that have been deployed to the serving environment, the
```
Comment (strickvl) [inline docs/book/stacks-and-components/component-guide/model-deployers/model-deployers.md:88]: See below, but wondering about this reference to the DB... we don't talk about the database elsewhere... why does a user have to know?

File: src/zenml/integrations/huggingface/services/huggingface_deployment.py [hunk 15]
```diff
@@ -162,6 +175,16 @@ def provision(self) -> None:
...
             )
 
+        # Check if the endpoint URL is available after provisioning
+        if hf_endpoint.url:
+            logger.info(
+                f"Hugging Face inference endpoint successfully deployed and available. Endpoint URL: {hf_endpoint.url}"
+            )
+        else:
+            logger.error(
+                "Failed to start Hugging Face inference endpoint service: No URL available."
```
Comment (strickvl) [inline src/zenml/integrations/huggingface/services/huggingface_deployment.py:185]: Is there any way we can get the user more information at this point? Just saying 'it failed' is maybe a bit frustrating. Even just a next step for them to check this or that would be good. Ideally, the nature of the failure...

File: src/zenml/integrations/huggingface/services/huggingface_deployment.py [hunk 16]
```diff
@@ -170,39 +193,30 @@ def check_status(self) -> Tuple[ServiceState, str]:
...
+                )
+            elif status == InferenceEndpointStatus.PENDING:
+                return (ServiceState.PENDING_STARTUP, "")
+            return (ServiceState.PENDING_STARTUP, "")
+        except (InferenceEndpointError, HfHubHTTPError):
             return (
-                ServiceState.PENDING_STARTUP,
-                "Hugging Face Inference Endpoint deployment is being created: ",
+                ServiceState.INACTIVE,
+                "Hugging Face Inference Endpoint deployment is inactive or not found",
```
Comment (strickvl) [inline src/zenml/integrations/huggingface/services/huggingface_deployment.py:218]: One of these ends in `: ` and the other has no space or colon. What is the correct formatting needed for these error messages? (I'm not sure where they get returned up into... are they printed to the screen somewhere?

File: src/zenml/integrations/seldon/steps/seldon_deployer.py [hunk 17]
```diff
@@ -73,13 +73,12 @@ def seldon_model_deployer_step(
...
     context = get_step_context()
     pipeline_name = context.pipeline.name
-    run_name = context.pipeline_run.name
     step_name = context.step_run.name
 
     # update the step configuration with the real pipeline runtime information
     service_config = service_config.copy()
     service_config.pipeline_name = pipeline_name
-    service_config.run_name = run_name
+    # service_config.run_name = run_name
```
Comment (strickvl) [inline src/zenml/integrations/seldon/steps/seldon_deployer.py:81]: remove this?

File: src/zenml/model_deployers/base_model_deployer.py [hunk 18]
```diff
@@ -226,9 +400,43 @@ def stop_model_server(
...
+                f"Failed to stop model server with UUID {uuid}: {e}"
+            ) from e
 
     @abstractmethod
+    def perform_start_model(
+        self,
+        service: BaseService,
+        timeout: int = DEFAULT_DEPLOYMENT_START_STOP_TIMEOUT,
+    ) -> BaseService:
+        """Abstract method to start a model server.
```
Comment (strickvl) [inline src/zenml/model_deployers/base_model_deployer.py:430]: server or service or model? Maybe nothing to be done about this, but the naming seems to use model, server and service a bit interchangeably...

File: src/zenml/model_deployers/base_model_deployer.py [hunk 19]
```diff
@@ -226,9 +400,43 @@ def stop_model_server(
...
+
+        Args:
+            service: The service to start.
```
Comment (strickvl) [inline src/zenml/model_deployers/base_model_deployer.py:433]: see down here you talk about a service, but the method name talks about a model, and then in the docstring you talk about a model server..

**Review by strickvl — CHANGES_REQUESTED** [review]
Love how much this simplifies things going forward! Nice work. Just some nits here and there, otherwise looking good. (CI is also failing).

File: src/zenml/services/service.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (avishniakov) [reply src/zenml/services/service.py:53]: This decorator takes pre_status in, what will happen if the service is not in that pre_status actually? Any impact?

File: src/zenml/model_deployers/base_model_deployer.py (hunk 18 above)
Comment (safoinme) [reply src/zenml/model_deployers/base_model_deployer.py:430]: That's a good point maybe we should unify the naming for all of these, I will start a thread in discord for this

File: docs/book/stacks-and-components/component-guide/model-deployers/mlflow.md [hunk 20]
```diff
@@ -119,9 +168,8 @@ from zenml.integrations.mlflow.services import MLFlowDeploymentService
 def prediction_service_loader(
     pipeline_name: str,
     pipeline_step_name: str,
-    running: bool = True,
     model_name: str = "model",
-) -> MLFlowDeploymentService:
+) -> None:
```
[the remark this answers is not in the document]
Comment (safoinme) [reply docs/book/stacks-and-components/component-guide/model-deployers/mlflow.md:172]: there is but works with a given exact service id and not with search

**Review by avishniakov — APPROVED** [review]
No more comments 🙂 
LGTM!

**Review by bcdurak — APPROVED** [review]

**Review by strickvl — APPROVED** [review]

**Pull request merged**